### DISCUSSION TITLE
Question regarding SCTransform, v2 and removing clusters

### DISCUSSION THREAD

**Discussion by pabloatria18** [satijalab/seurat#5521] [Q&A]
Hello!
I am trying to run the updated version of SCTransform on my dataset and I am wondering when is the right time in the workflow to remove undesired clusters (in my case hemato contamination) and re-do the clustering.

Thank you

**Comment by samuel-marsh** [comment]
Hi,

Not member of dev team but hopefully can be helpful.  In my opinion any time you remove clusters (or any cells) from analysis you should repeat the steps of the analysis that were based on the presence of those cells.  This definitely includes construction of nearest neighbor graphs as well as the dimensionality reduction.  But also steps identifying variables genes, running PCA etc.  Those were all based (in part) on past information (cells) that are no longer present and therefore don't necessarily accurately reflect the current makeup of the analysis.

Best,
Sam

**Reply by pabloatria18** [reply]
Hello Sam! Thank you for your answer.
I am doing it like this, but with SCT v2 It seems I can't remove the clusters because I get the same UMAP before and after removing,

Here is my code:

#load data
my.data.control <- Read10X(data.dir = "C:/Users/Pablo/Documents/sc2022/Unloaded")
my.data.loaded <- Read10X(data.dir = "C:/Users/Pablo/Documents/sc2022/Loaded")

#make seurat object
my.seurat1 <- CreateSeuratObject(counts = my.data.control, project = "loaded", min.cells = 3, min.features = 200)
my.seurat2 <- CreateSeuratObject(counts = my.data.loaded, project = "loaded", min.cells = 3, min.features = 200)

#FILTER
my.seurat1[["percent.mt"]] <- PercentageFeatureSet(my.seurat1, pattern = "^mt-")
my.seurat2[["percent.mt"]] <- PercentageFeatureSet(my.seurat2, pattern = "^mt-")

VlnPlot(my.seurat1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(my.seurat2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

#Subset to cells with > 200 genes, mt < 20%
my.seurat1 <- subset(my.seurat1, subset = nFeature_RNA > 500 & percent.mt < 20)
my.seurat2 <- subset(my.seurat2, subset = nFeature_RNA > 500 & percent.mt < 20)

# Name the samples 
group1 <- sample(c("Control"), size = 2601, replace = TRUE)
group2 <- sample(c("Loaded"), size = 3215, replace = TRUE)

names(group1) <- colnames(my.seurat1)
names(group2) <- colnames(my.seurat2)

ctrl <- AddMetaData(object = my.seurat1, metadata = group1, col.name = "group")
loaded <- AddMetaData(object = my.seurat2, metadata = group2, col.name = "group")

my.list <- list(Control = ctrl, Loaded = loaded)

ctrl <- SCTransform(ctrl, vst.flavor = "v2", verbose = FALSE) %>%
  RunPCA(npcs = 30, verbose = FALSE) %>%
  RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
  FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
  FindClusters(resolution = 0.7, verbose = FALSE)

DimPlot(ctrl, label = T, repel = T) + ggtitle("Unsupervised clustering")

loaded <- SCTransform(loaded, vst.flavor = "v2", verbose = FALSE) %>%
  RunPCA(npcs = 30, verbose = FALSE)
ifnb.list <- list(Control = ctrl, Loaded = loaded)
features <- SelectIntegrationFeatures(object.list = ifnb.list, nfeatures = 3000)
ifnb.list <- PrepSCTIntegration(object.list = ifnb.list, anchor.features = features)

group.anchors <- FindIntegrationAnchors(object.list = ifnb.list, normalization.method = "SCT",
                                        anchor.features = features)

combined.sct <- IntegrateData(anchorset = group.anchors, normalization.method = "SCT")

combined.sct <- RunPCA(combined.sct, verbose = FALSE)
combined.sct <- RunUMAP(combined.sct, reduction = "pca", dims = 1:20, verbose = FALSE)
combined.sct <- FindNeighbors(combined.sct, reduction = "pca", dims = 1:20)
combined.sct <- FindClusters(combined.sct, resolution = 0.3)

DimPlot(combined.sct, reduction = "umap", label = T)


## REMOVING CLUSTERS **( SOMEWHERE HERE IS THE ISSUE I GUESS)**
FeaturePlot(combined.sct, features = c("Ptprc", "Hbb-bt","Myl1","Myod1","Cdh5")) 

combined.sct <- subset(combined.sct, idents = c("3","2","1","11","8","14"), invert = TRUE)

split.list <- SplitObject(combined.sct, split.by = "group")
my.list <- split.list[c("ctrl", "loaded")]

group.anchors <- FindIntegrationAnchors(object.list = ifnb.list, normalization.method = "SCT",
                                        anchor.features = features)

combined.sct <- IntegrateData(anchorset = group.anchors, normalization.method = "SCT")

combined.sct <- RunPCA(combined.sct, verbose = FALSE)
combined.sct <- RunUMAP(combined.sct, reduction = "pca", dims = 1:20, verbose = FALSE)
combined.sct <- FindNeighbors(combined.sct, reduction = "pca", dims = 1:20)
combined.sct <- FindClusters(combined.sct, resolution = 0.3)

DimPlot(combined.sct, reduction = "umap", label = T)

**Reply by selenavana** [reply]
Hello,

Please let me know if you figure this out, as I am trying to do the same thing. I might suggest creating a new object for your subset, like:

subset.combinedsct <- subset(combined.sct, idents = c("3","2","1","11","8","14"), invert = TRUE)

and then perhaps run the PCA and UMAP off of object subset.combinedsct. I am not sure if this will work, but it's a possible starting point? 

warmly,
selena

**Reply by saketkc** [reply]
> Hello Sam! Thank you for your answer. I am doing it like this, but with SCT v2 It seems I can't remove the clusters because I get the same UMAP before and after removing,
> 
> Here is my code:
> 
> #load data my.data.control <- Read10X(data.dir = "C:/Users/Pablo/Documents/sc2022/Unloaded") my.data.loaded <- Read10X(data.dir = "C:/Users/Pablo/Documents/sc2022/Loaded")
> 
> #make seurat object my.seurat1 <- CreateSeuratObject(counts = my.data.control, project = "loaded", min.cells = 3, min.features = 200) my.seurat2 <- CreateSeuratObject(counts = my.data.loaded, project = "loaded", min.cells = 3, min.features = 200)
> 
> #FILTER my.seurat1[["percent.mt"]] <- PercentageFeatureSet(my.seurat1, pattern = "^mt-") my.seurat2[["percent.mt"]] <- PercentageFeatureSet(my.seurat2, pattern = "^mt-")
> 
> VlnPlot(my.seurat1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3) VlnPlot(my.seurat2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
> 
> #Subset to cells with > 200 genes, mt < 20% my.seurat1 <- subset(my.seurat1, subset = nFeature_RNA > 500 & percent.mt < 20) my.seurat2 <- subset(my.seurat2, subset = nFeature_RNA > 500 & percent.mt < 20)
> 
> # Name the samples
> group1 <- sample(c("Control"), size = 2601, replace = TRUE) group2 <- sample(c("Loaded"), size = 3215, replace = TRUE)
> 
> names(group1) <- colnames(my.seurat1) names(group2) <- colnames(my.seurat2)
> 
> ctrl <- AddMetaData(object = my.seurat1, metadata = group1, col.name = "group") loaded <- AddMetaData(object = my.seurat2, metadata = group2, col.name = "group")
> 
> my.list <- list(Control = ctrl, Loaded = loaded)
> 
> ctrl <- SCTransform(ctrl, vst.flavor = "v2", verbose = FALSE) %>% RunPCA(npcs = 30, verbose = FALSE) %>% RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE) %>% FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>% FindClusters(resolution = 0.7, verbose = FALSE)
> 
> DimPlot(ctrl, label = T, repel = T) + ggtitle("Unsupervised clustering")
> 
> loaded <- SCTransform(loaded, vst.flavor = "v2", verbose = FALSE) %>% RunPCA(npcs = 30, verbose = FALSE) ifnb.list <- list(Control = ctrl, Loaded = loaded) features <- SelectIntegrationFeatures(object.list = ifnb.list, nfeatures = 3000) ifnb.list <- PrepSCTIntegration(object.list = ifnb.list, anchor.features = features)
> 
> group.anchors <- FindIntegrationAnchors(object.list = ifnb.list, normalization.method = "SCT", anchor.features = features)
> 
> combined.sct <- IntegrateData(anchorset = group.anchors, normalization.method = "SCT")
> 
> combined.sct <- RunPCA(combined.sct, verbose = FALSE) combined.sct <- RunUMAP(combined.sct, reduction = "pca", dims = 1:20, verbose = FALSE) combined.sct <- FindNeighbors(combined.sct, reduction = "pca", dims = 1:20) combined.sct <- FindClusters(combined.sct, resolution = 0.3)
> 
> DimPlot(combined.sct, reduction = "umap", label = T)
> 
> ## REMOVING CLUSTERS **( SOMEWHERE HERE IS THE ISSUE I GUESS)**
> FeaturePlot(combined.sct, features = c("Ptprc", "Hbb-bt","Myl1","Myod1","Cdh5"))
> 
> combined.sct <- subset(combined.sct, idents = c("3","2","1","11","8","14"), invert = TRUE)
> 
> split.list <- SplitObject(combined.sct, split.by = "group") my.list <- split.list[c("ctrl", "loaded")]
> 
> group.anchors <- FindIntegrationAnchors(object.list = ifnb.list, normalization.method = "SCT", anchor.features = features)
> 
> combined.sct <- IntegrateData(anchorset = group.anchors, normalization.method = "SCT")
> 
> combined.sct <- RunPCA(combined.sct, verbose = FALSE) combined.sct <- RunUMAP(combined.sct, reduction = "pca", dims = 1:20, verbose = FALSE) combined.sct <- FindNeighbors(combined.sct, reduction = "pca", dims = 1:20) combined.sct <- FindClusters(combined.sct, resolution = 0.3)
> 
> DimPlot(combined.sct, reduction = "umap", label = T)

After you remove the clusters you should ideally use only these cells to perform the rest of the steps in the workflow. Which means, your `my.list` should be treated as `ifnb.list` in the beginning of the workflow and all the steps need to be re-run.
